verification/src/solidity/verifier.rs

```diff
@@ -1,51 +1,281 @@
+                        }
+                    }
+                }
+                Ok(_) => {
+                    // if key is not "solc" str we may skip the corresponding value
+                    d.skip()?;
+                }
+                Err(err) if err.is_type_mismatch() => {
+                    // if key is not `str` we may skip the corresponding value
+                    d.skip()?;
+                }
+                Err(err) => return Err(err),
+            }
+        }
+
+        // We require that no elements left in the decoder when
+        // the whole map has been processed. That adds another layer
+        // of assurance that encoded bytes are actually metadata hash.
+        if d.datatype().is_ok() {
+            return Err(Error::custom(ParseMetadataHashError::NonExhausted));
+        }
+
+        let solc = solc.map(bytes::Bytes::copy_from_slice);
+        Ok(MetadataHash { solc })
+    }
+
+    fn nil() -> Option<Self> {
+        Some(Self { solc: None })
+    }
+}
+
 /// Wrapper under `evm.deployedBytecode` from the standard output JSON
 /// (https://docs.soliditylang.org/en/latest/using-the-compiler.html#output-description).
 ///
 /// Provides an interface to retrieve parts the deployed bytecode consists of:
-/// actual bytecode participating in EVM transaction execution and optionally metadata hash.
+/// actual bytecode participating in EVM transaction execution and metadata hash.
 #[derive(Clone, Debug, PartialEq)]
-struct DeployedBytecode {}
+struct DeployedBytecode {
+    /// Bytecode without metadata hash
+    bytecode: bytes::Bytes,
+    /// Metadata hash encoded into bytecode
+    metadata_hash: MetadataHash,
+    /// Raw deployed bytecode bytes
+    bytes: bytes::Bytes,
+}
+
+impl DeployedBytecode {
+    /// Returns deployed bytecode without metadata hash
+    pub fn bytecode(&self) -> bytes::Bytes {
+        self.bytecode.clone()
+    }
+
+    /// Returns a metadata hash
+    pub fn metadata_hash(&self) -> &MetadataHash {
+        &self.metadata_hash
+    }
+
+    /// Returns metadata hash encoded as bytes and concatenated with 2 bytes representing its length
+    pub fn encoded_metadata_hash_with_length(&self) -> bytes::Bytes {
+        let start = self.bytecode.len();
+        let end = self.bytes.len();
+        self.bytes.slice(start..end)
+    }
+}
 
 impl FromStr for DeployedBytecode {
     type Err = InitializationError;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        todo!()
+        let bytes = DisplayBytes::from_str(s)
+            .map_err(|_| InitializationError::InvalidDeployedBytecode)?
+            .0;
+
+        DeployedBytecode::try_from(bytes)
+    }
+}
+
+impl TryFrom<bytes::Bytes> for DeployedBytecode {
+    type Error = InitializationError;
+
+    fn try_from(encoded: bytes::Bytes) -> Result<Self, Self::Error> {
+        // If metadata is present, last two bytes encode its length in a two-byte big-endian encoding
+        if encoded.len() < 2 {
```

Comment:
I suggest adding a struct, which will declaratively describe the bytecode format, i.e.
```rust
struct Bytecode {
  // bytecode starts with this
  compiler: Compiler,
  // 4 zeroes specifying the next value
  guard: Guard,
  // something something size of 16 bytes
  foo: Foo,
  // last 2 bytes are the length
  length: u16,
}
```

so you could just look at it and understand what format bytecode is, instead of looking at the `parse` function and trying to guess something from it (like reading here, that last two bytes are the length etc)
and then implement parse on it (may be `DeployedBytecode` already is this struct, but then I'd like to see all the values)

_In C, one of the important usages of struct was memory mapping, so you could just convert a pointer to the raw memory (like &[u8]) to this struct and get all the correct values. Time had passed, there is a lot of nuance nowadays (like different endiannesses), but I think the idea in general to "declare" the format for the part of the memory (or bytes) is kinda nice._

Reply:
I believe it is an unnecessary work and information right now. as It is an internal structure which is not exposed to outside the module. Current layout allows to effectively and conveniently operate with the structure internally and provide some public api for the users of that structure.

Reply:
that's not for public interface, but for other developers. With json, for example, you can just look at the json itself to understand the format. With binary formats you can't do that, so it is very helpful if you can see the format "specification" somewhere. Think about yourself after 3 months or about another developer, when you/he'll need to add/fix something connected to the binary format

Reply:
it's like a comment explaining the format, but more strict

Reply:
we can do this in another PR tho

Reply:
I'll create an issue to consider that question after PR will be merged